csharp/src/Drivers/Apache/Hive2/HiveServer2TlsImpl.cs

```diff
@@ -155,14 +180,39 @@ static internal bool ValidateCertificate(X509Certificate? cert, SslPolicyErrors
                 return !policyErrors.HasFlag(SslPolicyErrors.RemoteCertificateChainErrors) || (tlsProperties.AllowSelfSigned && IsSelfSigned(cert2));
             }
 
-            X509Certificate2 trustedRoot = new X509Certificate2(tlsProperties.TrustedCertificatePath);
             X509Chain customChain = new();
-            customChain.ChainPolicy.ExtraStore.Add(trustedRoot);
             // "tell the X509Chain class that I do trust this root certs and it should check just the certs in the chain and nothing else"
             customChain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
+            var collection = LoadPemCertificates(tlsProperties.TrustedCertificatePath!);
+
+            foreach (var trustedCert in collection)
+            {
+                customChain.ChainPolicy.ExtraStore.Add(trustedCert);
+            }
             customChain.ChainPolicy.RevocationMode = X509RevocationMode.Online;
 
             bool chainValid = customChain.Build(cert2);
+            if (chainValid)
+            {
+                bool trustedBy = false;
+                foreach (X509ChainElement element in customChain.ChainElements)
+                {
+                    foreach (X509Certificate2 ca in collection)
+                    {
+                        if (element.Certificate.Thumbprint == ca.Thumbprint)
+                        {
+                            trustedBy = true;
+                            break;
+                        }
+                    }
+                    if (trustedBy)
+                    {
+                        break;
+                    }
+                }
+                chainValid = chainValid && trustedBy;
+            }
+
             return chainValid || (tlsProperties.AllowSelfSigned && IsSelfSigned(cert2));
```

Comment:
Consider modifying the check around line 178 to return early for a self-signed certificate even if `TrustedCertificatePath` is not null. There's no reason to go to the trouble of loading the additional certificates if we already know we have a self-signed cert. Then the check here becomes superfluous and can be removed.